Index: CraftBukkit/src/main/java/net/minecraft/server/Explosion.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- CraftBukkit/src/main/java/net/minecraft/server/Explosion.java	(date 1429830155000)
+++ CraftBukkit/src/main/java/net/minecraft/server/Explosion.java	(revision )
@@ -266,7 +266,7 @@
     }
 
     public EntityLiving c() {
-        return this.source == null ? null : (this.source instanceof EntityTNTPrimed ? ((EntityTNTPrimed) this.source).getSource() : (this.source instanceof EntityLiving ? (EntityLiving) this.source : null));
+        return this.source == null ? null : (this.source instanceof EntityTNTPrimed ? ((EntityTNTPrimed) this.source).getSource() : (this.source instanceof EntityLiving ? (EntityLiving) this.source : (this.source instanceof EntityFireball ? ((EntityFireball) this.source).shooter : null)));
     }
 
     public void clearBlocks() {
